llm/src/flash_attention.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Triton implementation of Flash Attention Layer
+
+Modified from:
+https://raw.githubusercontent.com/HazyResearch/flash-attention/main/flash_attn/flash_attention.py
+"""
+
+import math
+import torch
+import torch.nn as nn
+from torch import Tensor
+from einops import rearrange
+from typing import List, Optional, Tuple, Union
+
+try:
+    from src.flash_attn_triton import flash_attn_qkvpacked_func, flash_attn_func
+except ImportError as e:
+    raise e
+
+
+class FlashAttention(nn.Module):
+    """Implement the scaled dot product attention with softmax.
+    Arguments
+    ---------
+        softmax_scale: The temperature to use for the softmax attention.
+                      (default: 1/sqrt(d_keys) where d_keys is computed at
+                      runtime)
+        attention_dropout: The dropout rate to apply to the attention
+                           (default: 0.0)
+    """
+    def __init__(self, num_heads, softmax_scale=None, attention_dropout=0.0, device=None, dtype=None):
+        super().__init__()
+        self.num_heads = num_heads
+        self.softmax_scale = softmax_scale
+        assert attention_dropout == 0
```

Comment:
This entire file is a copy pasta from [hazy research](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attention.py); then modify it to use the triton kernel instead of the cuda flash attn kernel.
I kept it to keep the module signature the same, but to your point, I should probably just remove it and move the assertion higher in the stack.